src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/Serialization/StringCache.cs

```diff
@@ -0,0 +1,162 @@
+                {
+                    return result!;
+                }
+                else
+                {
+                    var weakRef = new Entry(key);
+                    _hashSet.Add(weakRef);
+
+                    // Whenever we expand lets clean up dead references
+                    if (_capacity <= _hashSet.Count)
+                    {
+                        Cleanup();
+                        _capacity *= 2;
+                    }
+
+                    return key;
+                }
+            }
+        }
+
+        private void Cleanup()
+        {
+            _hashSet.RemoveWhere((weakRef) =>
+            {
+                return !weakRef.IsAlive;
+            });
+        }
+
+        private bool TryGetValue(string key, out string? value)
+        {
+            if (_hashSet.Contains(new Entry(key)))
+            {
+                var exfiltrator = (ExfiltratingEqualityComparer)_hashSet.Comparer;
+                var val = exfiltrator.LastEqualValue!;
+                if (val.TryGetTarget(out var target))
+                {
+                    value = target!;
+                    return true;
+                }
+                else
+                {
+                    throw new InvalidOperationException("HashSet contained entry but we were unable to get the value from it. This should be impossible");
+                }
+            }
+
+            value = default;
+            return false;
+        }
+
+        /// <summary>
+        /// This is a gross hack to do a sneaky and get the value inside the HashSet out given the lack of any Get operations in netstandard2.0.
+        /// If we ever upgrade to 2.1 delete this and just use the built in TryGetValue method.
+        /// </summary>
+        /// <remarks>
+        /// This is fragile on the ordering of the values passed to the EqualityComparer by HashSet.
+        /// If that ever switches we have to react, if it becomes indeterminate we have to abandon this strategy.
+        /// </remarks>
+        private class ExfiltratingEqualityComparer : IEqualityComparer<Entry>
+        {
+            public Entry? LastEqualValue { get; private set; }
+
+            public bool Equals(Entry x, Entry y)
+            {
+                if (x.Equals(y))
+                {
+                    LastEqualValue = x;
+                    return true;
+                }
+                else
+                {
+                    LastEqualValue = null;
+                    return false;
+                }
+            }
+
+            public int GetHashCode(Entry obj)
+            {
+                return obj.TargetHashCode;
+            }
+        }
+
+        private class Entry
+        {
+            // In order to use HashSet we need a stable HashCode, so we have to cache it as soon as it comes in.
+            // If the HashCode is unstable then entries in the HashSet become unreachable/unremovable.
+            public readonly int TargetHashCode;
```

Comment:
Why public? Couldn't ExfiltratingEqualityComparer just use GetHashCode() instead of this?